src/engine/plugins/match.cpp

```diff
@@ -193,6 +169,36 @@ Status MatchPlugin::HandleRequest(const RoutingAlgorithmsInterface &algorithms,
 
     auto candidates_lists = GetPhantomNodesInRange(facade, tidied.parameters, search_radiuses);
 
+    // Add previous states for stand_in_place behavior
+    for (const auto current_index :
```

Comment:
I think you could get rid of this step and it would still include `PhantomNodes` from the previous step. They key the changed `IsIndistinct, that not only checks for ID but also for the offset. So if there are segments that are probable for both candidates they should be included in the result from both RTree queries.

Reply:
I think I did not fully understand your idea.
How do we obtain the PhantomNodes representing the previous states?

I am talking about this situation: lets suppose we have a noisy GPS signal ABCD along the single edge of the road graph. 
```
A______B
     C<______D
================
Z    X Y     W   
```
If the entire path is one edge, and the distance between the points is large enough, then `GetPhantomNodesInRange` returns the states of Z Y X W. To visit these states, the algorithm will have to add to the route the movement in the opposite direction Y-X, and in the resulting geometry this section of the road will be processed 3 times. My idea is that we need to add neighboring states. So that the point C can lead to the state Y. Then the resulting sequence on the edge becomes ZYYW. This addition is the responsibility of this code.

Reply:
> If the entire path is one edge

Ah okay yes that is indeed a problem, I did not think of this edge case - I think we need to rework some of the modeling to make this actually work algorithmically.

Right now we represent states as `(edge_based_node_id, segment_id, weight_offset)`. The problem with this representation is that the `weight_offset` is a continuous number so there are infinitely many states per edge segment that we could encounter. In your example above this would look something like this:

Z : (0, 0, 0.0)
X : (0, 0, 0.45)
Y : (0, 0, 0.5)
W: (0, 0, 0.9)

For the input coordinate `C` the R-Tree will only give us the state `X` since its the nearest point on the edge segment. The main problem here is that we would never consider `Y` to be the "true" origin of `C`, hence leading to a very low transition probability from `B` to `C`.

To work around this problem we could try to make the states discrete. If we have a long edge segment, we could try to split it into small `sub-segments` and only snap the candidate states to these. Then for every candidate we find using the R-Tree query, we don't just add all candidate segments, but also all sub-segments in range.

1. Normal R-Tree query right now will identify all segments in the search radius.
2. For every candidate segment, compute the segment length:
    - If the segment length is below the threshold snap to one of the endpoints.
    - If the segment length is above the threshold, sample points along the segment and insert each as candidate.

By making this discrete, the candidates for segment coordinate `C` would include all candidates for segment `B` that are in query distance.

I need to think about this a little more but I think you found one big modeling problem, you would run into when using very high-precision traces.